zephyr/lib/alloc.c

```diff
@@ -68,12 +68,109 @@ extern char _end[], _heap_sentry[];
 
 static struct k_heap sof_heap;
 
+#if CONFIG_L3_HEAP
+static struct k_heap l3_heap;
+
+/**
+ * Returns the start of L3 memory heap.
+ * @return Pointer to the L3 memory location which can be used for L3 heap.
+ */
+static inline uintptr_t get_l3_heap_start(void)
+{
+	/*
+	 * TODO: parse the actual offset using:
+	 * - HfIMRIA1 register
+	 * - rom_ext_load_offset
+	 * - main_fw_load_offset
+	 * - main fw size in manifest
+	 */
+
+	return (uintptr_t)ROUND_UP(IMR_L3_HEAP_BASE, L3_MEM_PAGE_SIZE);
+}
+
+/**
+ * Returns the size of L3 memory heap.
+ * @return Size of the L3 memory region which can be used for L3 heap.
+ */
+static inline size_t get_l3_heap_size(void)
+{
+	 /*
+	  * Calculate the IMR heap size using:
+	  * - total IMR size
+	  * - IMR base address
+	  * - actual IMR heap start
+	  */
+	return ROUND_DOWN(IMR_L3_HEAP_SIZE, L3_MEM_PAGE_SIZE);
+}
+
+/**
+ * Checks whether pointer is from L3 heap memory range.
+ * @param ptr Pointer to memory being checked.
+ * @return True if pointer falls into L3 heap region, false otherwise.
+ */
+static bool is_l3_heap_pointer(void *ptr)
+{
+	uintptr_t l3_heap_start = get_l3_heap_start();
+	uintptr_t l3_heap_end = l3_heap_start + get_l3_heap_size();
+
+	if (POINTER_TO_UINT(ptr) >= l3_heap_start &&
+	    POINTER_TO_UINT(ptr) < l3_heap_end) {
+		return true;
+	}
+	return false;
+}
+
+/**
+ * Allocate buffer from the L3 memory heap.
+ * Return pointer is always aligned to page size.
+ * @param size - Size of memory to be allocated
+ * @return Pointer to allocated L3 memory block.
+ */
+void *l3_alloc(size_t size)
+{
+	k_spinlock_key_t key;
+	void *ret;
+
+	key = k_spin_lock(&l3_heap.lock);
+	ret = sys_heap_aligned_alloc(&l3_heap.heap, L3_MEM_PAGE_SIZE, size);
+	k_spin_unlock(&l3_heap.lock, key);
+	return ret;
+}
+
+/**
+ * Allocate buffer from the L3 memory heap.
+ * @param mem SPointer to memory block to be freed.
+ * @return Pointer to allocated L3 memory block.
+ */
+void l3_free(void *mem)
+{
+	/*
+	 * first check to which heap the mem argument belongs to
+	 */
+	uintptr_t l3_heap_start = get_l3_heap_start();
+	uintptr_t l3_heap_end = l3_heap_start + get_l3_heap_size();
+
+	if ((uintptr_t)mem < l3_heap_start && (uintptr_t)mem >= l3_heap_end)
+		return;
+
+	k_spinlock_key_t key = k_spin_lock(&l3_heap.lock);
+
+	sys_heap_free(&l3_heap.heap, mem);
```

Comment:
are these allocations always going to be uncached?

Reply:
good point! it is always CACHED (adress area 0xA1...)

good that currently IMR heap is used exclusively by core 0
Anyway "cache resistance" as in HPSRAM heap is needed here 

ANWYAY, for now I can do a workaround for this
 - an assert limiting alloc/free of IMR heap to core 0
 - open a defect/feture req for multicore implementation

we do need IMR heap for D3 a.k.a "imr boot" flow, badly needed for chrome team :)

Reply:
fixed, handling of l3 heap redesigned